server/src/main/java/org/apache/druid/metadata/segment/SqlSegmentsMetadataManagerV2.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.druid.metadata.segment.cache.SegmentMetadataCache;
+import org.apache.druid.segment.metadata.CentralizedDatasourceSchemaConfig;
+import org.apache.druid.segment.metadata.SegmentSchemaCache;
+import org.apache.druid.server.coordinator.DruidCompactionConfig;
+
+/**
+ * Implementation V2 of {@link SegmentsMetadataManager}, that can use the
+ * segments cached in {@link SegmentMetadataCache} to build a {@link DataSourcesSnapshot}.
+ * <p>
+ * This class acts as a wrapper over {@link SqlSegmentsMetadataManager} and the
+ * {@link SegmentMetadataCache}. If the cache is enabled, an additional poll is
+ * not done and the segments already present in the cache are used to build the
+ * snapshot. If the {@link SegmentMetadataCache} is disabled, the polling is
+ * delegated to the legacy implementation in {@link SqlSegmentsMetadataManager}.
+ * <p>
+ * The Coordinator always uses the snapshot to perform various segment management
+ * duties such as loading, balancing, etc.
+ * The Overlord uses the snapshot only when compaction supervisors are enabled.
+ * Thus, when running the Overlord as a standalone service (i.e. not combined
+ * with the Coordinator), {@link #startPollingDatabasePeriodically()} and
+ * {@link #stopPollingDatabasePeriodically()} are called based on the current
+ * state of {@link DruidCompactionConfig#isUseSupervisors()}.
+ */
+@ManageLifecycle
+public class SqlSegmentsMetadataManagerV2 implements SegmentsMetadataManager
+{
+  private static final Logger log = new Logger(SqlSegmentsMetadataManagerV2.class);
+
+  private final SegmentsMetadataManager delegate;
+  private final SegmentMetadataCache segmentMetadataCache;
+  private final SegmentsMetadataManagerConfig managerConfig;
+  private final CentralizedDatasourceSchemaConfig schemaConfig;
+
+  public SqlSegmentsMetadataManagerV2(
+      SegmentMetadataCache segmentMetadataCache,
+      SegmentSchemaCache segmentSchemaCache,
+      SQLMetadataConnector connector,
+      Supplier<SegmentsMetadataManagerConfig> managerConfig,
+      Supplier<MetadataStorageTablesConfig> tablesConfig,
+      CentralizedDatasourceSchemaConfig centralizedDatasourceSchemaConfig,
+      ServiceEmitter serviceEmitter,
+      ObjectMapper jsonMapper
+  )
+  {
+    this.delegate = new SqlSegmentsMetadataManager(
+        jsonMapper,
+        managerConfig, tablesConfig, connector, segmentSchemaCache,
+        centralizedDatasourceSchemaConfig, serviceEmitter
+    );
+    this.managerConfig = managerConfig.get();
+    this.segmentMetadataCache = segmentMetadataCache;
+    this.schemaConfig = centralizedDatasourceSchemaConfig;
+  }
+
+  /**
+   * @return true if segment metadata cache is enabled and segment schema cache
+   * is not enabled. Segment metadata cache currently does not handle segment
+   * schema updates.
+   */
+  private boolean useCacheToBuildTimeline()
+  {
+    return segmentMetadataCache.isEnabled() && !schemaConfig.isEnabled();
```

Comment:
I guess the `!schemaConfig.isEnabled()` escape hatch is here to maintain compatibility with `druid.centralizedDatasourceSchema.enabled = true`. Otherwise the schemas would not be polled.

IMO, in the short term, it should be an error to specify `druid.centralizedDatasourceSchema.enabled = true` when the incremental cache is enabled. Otherwise people will potentially enable both features and wonder why the incremental cache isn't working.

In the long term, how would these two features coexist?

Reply:
Thanks for the suggestion! It makes sense to error out if both features are enabled simultaneously, at least for the time being. 

I wanted to address handling both features being enabled in this PR itself but since the patch became a little bloated, I decided to do it in a follow up.

The long term solution would be to also poll the schemas inside the `HeapMemorySegmentMetadataCache` if schema caching is enabled.
Eventually, we will get rid of `SqlSegmentsMetadataManager` completely.